exiftool.go

```diff
@@ -4,53 +4,96 @@ import (
 	"bufio"
 	"bytes"
 	"encoding/json"
-	"errors"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"sync"
-)
 
-var binary = "exiftool"
-var executeArg = "-execute"
-var initArgs = []string{"-stay_open", "True", "-@", "-", "-common_args"}
-var extractArgs = []string{"-j"}
-var closeArgs = []string{"-stay_open", "False", executeArg}
-var readyToken = []byte("{ready}\n")
-var readyTokenLen = len(readyToken)
+	"github.com/pkg/errors"
+)
 
 // ErrNotExist is a sentinel error for non existing file
 var ErrNotExist = errors.New("file does not exist")
 
+// Config contains the configuration used by the module
+type Config struct {
+	binary        string
+	executeArg    string
+	initArgs      []string
+	extractArgs   []string
+	closeArgs     []string
+	readyToken    []byte
+	readyTokenLen int
+}
+
+// NewExiftoolConfig creates a new default configuration to be used with Exiftool. Pass configuration functions
+// to change the default configuration
+func NewExiftoolConfig(opts ...func(*Config) error) (*Config, error) {
+	config := Config{
+		binary:        "exiftool",
+		executeArg:    "-execute",
+		initArgs:      []string{"-stay_open", "True", "-@", "-", "-common_args"},
+		extractArgs:   []string{"-j"},
+		closeArgs:     nil,
+		readyToken:    []byte("{ready}\n"),
+		readyTokenLen: 0,
+	}
+
+	config.closeArgs = []string{"-stay_open", "False", config.executeArg}
+	config.readyTokenLen = len(config.readyToken)
+
+	for _, opt := range opts {
+		if err := opt(&config); err != nil {
+			return nil, errors.Wrap(err, "Error setting configuration options")
+		}
+	}
+
+	return &config, nil
+}
+
 // Exiftool is the exiftool utility wrapper
 type Exiftool struct {
+	config  *Config
 	lock    sync.Mutex
 	stdin   io.WriteCloser
 	stdout  io.ReadCloser
 	scanout *bufio.Scanner
 }
 
 // NewExiftool instanciates a new Exiftool with configuration functions. If anything went
-// wrong, a non empty error will be returned
-func NewExiftool(opts ...func(*Exiftool) error) (*Exiftool, error) {
+// wrong, a non empty error will be returned. If config is nil, it will use a default config
+func NewExiftool(config *Config, opts ...func(*Exiftool) error) (*Exiftool, error) {
```

Comment:
This change brakes API backward compatibility.
Moreover, the behaviour of the `ExtractMetadata` functiondeeply depends on the arguments provided to the `exiftool` tool.
Adding a "configuration" argument to the `NewExiftool` function might let user think that it can be customized but it's quite dangerous if the `ExtractMetadata` function is not adapted.
Garthering all constants within a struct is perhaps a good practice, but it shouldn'y be provided as argument.

Reply:
Agreed! I changed it now so it's not an argument, and the user must create a opt func if they wish to change the config, much like what you had already done. Like configuring the other options you had previously, a user must know what they are doing if they attempt to configure the parameters. By default, the NewExiftoolConfiguration leaves any non changed fields to their default values. I also unexported the config struct, to only allow instatiation by using the NewExiftoolConfig from outside the package. This will prevent unset values, and will make sure default, sane values are assigned unless the user explicitly changes them in a config function